pkg/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -904,9 +913,148 @@ func (q *DBJobQueue) jobDependents(ctx context.Context, conn connection, id uuid
 	return dependents, nil
 }
 
+// deleteJob removes the job from the database
+// the CASCADE constraint will also delete any entries from the job_dependencies table
+func (q *DBJobQueue) deleteJob(ctx context.Context, jobID uuid.UUID) error {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	_, err = conn.Exec(ctx, sqlDelete, jobID)
+	if err != nil {
+		q.logger.Error(err, "Error deleting job")
+		return err
+	}
+	return nil
+}
+
+// deleteJobDependencies removes job dependencies
+func (q *DBJobQueue) deleteJobDependencies(ctx context.Context, jobID, dependencyID uuid.UUID) error {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	_, err = conn.Exec(ctx, sqlDeleteDependencies, jobID, dependencyID)
+	if err != nil {
+		q.logger.Error(err, "Error deleting dependencies")
+		return err
+	}
+	return nil
+}
+
+// listJobs returns a list of all of the job UUIDs
+func (q *DBJobQueue) listJobs(ctx context.Context) (jobs []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	rows, err := conn.Query(ctx, sqlQueryListJobs)
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var t uuid.UUID
+		err = rows.Scan(&t)
+		if err != nil {
+			// Log the error and try to continue with the next row
+			q.logger.Error(err, "Unable to read job uuid from jobs")
+			continue
+		}
+		jobs = append(jobs, t)
+	}
+	if rows.Err() != nil {
+		q.logger.Error(rows.Err(), "Error reading job uuids from jobs")
+	}
+
+	return
+}
+
 // AllRootJobIDs returns a list of top level job UUIDs that the worker knows about
-func (q *DBJobQueue) AllRootJobIDs() ([]uuid.UUID, error) {
-	// TODO write this
+func (q *DBJobQueue) AllRootJobIDs(ctx context.Context) (rootJobs []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	var jobs []uuid.UUID
+	jobs, err = q.listJobs(ctx)
+	if err != nil {
+		return
+	}
+
+	for _, id := range jobs {
+		var dependents []uuid.UUID
+		dependents, err = q.jobDependents(ctx, conn, id)
+		if err != nil {
+			return
+		}
+
+		if len(dependents) == 0 {
+			rootJobs = append(rootJobs, id)
+		}
+	}
+
+	return
+}
+
+// DeleteJob deletes a job and all of its dependencies from the database
+// If a dependency has multiple dependents it will only remove the parent job from
+// the dependents list for that job instead of removing it.
+//
+// This assumes that the jobs have been created correctly, and that they have
+// no dependency loops. Shared Dependents are ok, but a job cannot have a dependency
+// on any of its parents (this should never happen).
+func (q *DBJobQueue) DeleteJob(ctx context.Context, id uuid.UUID) error {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	// Start it off with an empty parent
+	return q.deleteJobs(ctx, conn, uuid.UUID{}, id)
+}
```

Comment:
Consider creating an explicit transaction here just like in `Enqueue`.

Reply:
I don't think I can. The functions all take connections or acquire their own from the pool, etc. But I agree that it needs something to lock access. Delete won't run unless the job is finished, so the only thing to worry about is other delete operations so a lock should work.

Reply:
How is the mutex going to break? You'd have to call DeleteJob from inside DeleteJob to double lock it. I admit it is a bit of a hack, but so is making complex new interface type.

Reply:
If someone knows how to do this with pure SQL I'd be open to that. My SQL-foo is really rusty.

Reply:
I am not sure what exactly do you mean by pure SQL, all I am asking is to avoid mutex and just use a regular SQL transaction across multiple helper functions. Feel free to ignore, I can do a follow-up.

Reply:
So it ends up there is already an interface for using conn or tx but it was buried in the middle of the code. I added Exec to it and I think this new version is one you'll like. Updated PR incoming...